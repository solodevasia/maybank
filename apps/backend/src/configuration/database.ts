import { TypeOrmModuleOptions } from '@nestjs/typeorm';

export default {
  type: process.env.DB_TYPE,
  host: process.env.DB_HOST,
  port: process.env.DB_PORT,
  username: process.env.DB_USER,
  password: process.env.DB_PASSWORD,
  database: process.env.DB_NAME,
  synchronize: true,
  autoLoadEntities: true,
} as Partial<TypeOrmModuleOptions>;
