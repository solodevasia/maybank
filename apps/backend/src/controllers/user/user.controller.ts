import AuthGuard from '@maybank/decorators/auth-guard';
import {
  LoginField,
  QueryUserList,
  UpdatedSchema,
  UserRegisterField,
} from '@maybank/dto/user.dto';
import UserRepository from '@maybank/repository/user/user.repository';
import UserService from '@maybank/services/user/user.service';
import { CustomRequest } from '@maybank/types/request';
import {
  Body,
  Controller,
  Delete,
  Get,
  HttpCode,
  HttpStatus,
  Param,
  Post,
  Put,
  Query,
  Req,
  UseGuards,
} from '@nestjs/common';

@Controller('user')
export default class UserController {
  constructor(
    private readonly repository: UserRepository,
    private readonly service: UserService,
  ) {}

  @Post()
  @HttpCode(HttpStatus.CREATED)
  created(@Body() body: UserRegisterField) {
    return this.service.created(body);
  }

  @Post('/login/access')
  @HttpCode(HttpStatus.OK)
  loginAccess(@Body() body: LoginField) {
    return this.service.login(body);
  }

  @Get('/profile')
  @HttpCode(HttpStatus.OK)
  @UseGuards(AuthGuard)
  profile(@Req() req: CustomRequest) {
    return this.repository.findOne(req.user.id);
  }

  @Put('/:id')
  @HttpCode(HttpStatus.OK)
  @UseGuards(AuthGuard)
  updated(@Param('id') id: number, @Body() body: UpdatedSchema) {
    return this.service.updated(id, body);
  }

  @Delete('/:id')
  @UseGuards(AuthGuard)
  @HttpCode(HttpStatus.OK)
  delete(@Param('id') id: number, @Req() req: CustomRequest) {
    return this.service.destroy(id, req.user.role);
  }

  @Get()
  @HttpCode(HttpStatus.OK)
  @UseGuards(AuthGuard)
  list(@Query() query: QueryUserList) {
    return this.repository.find(query);
  }

  @Get('/:id')
  @HttpCode(HttpStatus.OK)
  @UseGuards(AuthGuard)
  detail(@Param('id') id: number) {
    return this.repository.findOne(id);
  }

  @Post('/logout')
  @UseGuards(AuthGuard)
  @HttpCode(HttpStatus.OK)
  logout(@Req() req: CustomRequest) {
    return this.service.logout(req.user.id);
  }
}
