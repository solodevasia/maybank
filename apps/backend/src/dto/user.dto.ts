import { Role } from '@maybank/types/role';
import { z, ZodIssueCode } from 'zod';

export const queryUserList = z.object({
  name: z.string().optional(),
  email: z.string().optional(),
  pic: z.string().optional(),
  role: z.number().default(0),
  page: z.number().default(1),
  limit: z.number().default(10),
});

export type QueryUserList = z.infer<typeof queryUserList>;

export const userRegisterField = z
  .object({
    name: z.string(),
    email: z.string().email(),
    password: z.string(),
    confirmation: z.string().optional(),
  })
  .superRefine((data, ctx) => {
    if (data.password !== data.confirmation) {
      ctx.addIssue({
        code: ZodIssueCode.custom,
        message: "Password don't match, please check again",
      });
    }
  });

export type UserRegisterField = z.infer<typeof userRegisterField>;

export const loginField = z.object({
  token: z.string(),
  password: z.string(),
});

export type LoginField = z.infer<typeof loginField>;

export const updatedSchema = z.object({
  name: z.string(),
  email: z.string(),
  role: z.enum(['member','staff','admin']),
  pic: z.string()
})


export type UpdatedSchema = z.infer<typeof updatedSchema>