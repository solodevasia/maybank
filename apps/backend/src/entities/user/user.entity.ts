import { BeforeInsert, Column, Entity, PrimaryGeneratedColumn } from 'typeorm';
import bcrypt from 'bcrypt';
import { Role } from '@maybank/types/role';

@Entity({
  name: 'user',
  schema: 'authenticates',
})
export default class UserEntity {
  @PrimaryGeneratedColumn()
  id!: number;

  @Column({ unique: true, nullable: false })
  name!: string;

  @Column({ unique: true, nullable: false })
  email!: string;

  @Column()
  password!: string;

  @Column({ nullable: true })
  pic!: string;

  @Column({ type: 'simple-enum', enum: Role, default: Role.user })
  role!: number;

  @Column()
  created_at: Date = new Date();

  @Column()
  is_active: boolean = false;

  @BeforeInsert()
  beforeInsert() {
    this.password = bcrypt.hashSync(this.password, 10);
    if (!this.role) {
      this.role = Role.user;
    }
  }

  checkPassword(password: string, hash: string) {
    return bcrypt.compareSync(password, hash);
  }
}
