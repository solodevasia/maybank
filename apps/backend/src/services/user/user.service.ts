import {
  LoginField,
  UpdatedSchema,
  userRegisterField,
  UserRegisterField,
} from '@maybank/dto/user.dto';
import UserEntity from '@maybank/entities/user/user.entity';
import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { JwtService } from '@nestjs/jwt';
import fs from 'fs';
import { join } from 'path';
import { Role } from '@maybank/types/role';

@Injectable()
export default class UserService {
  constructor(
    @InjectRepository(UserEntity)
    private readonly repository: Repository<UserEntity>,
    private readonly jwtService: JwtService,
  ) {}

  async created(body: UserRegisterField) {
    const parse = userRegisterField.safeParse(body);
    if (parse.error) {
      throw new HttpException(
        {
          message: parse.error.errors[0].message,
          status: HttpStatus.BAD_REQUEST,
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    const findOne = await this.repository.findOne({
      where: [{ name: body.name }, { email: body.email }],
    });
    if (findOne) {
      throw new HttpException(
        {
          message: 'Username or email already exists',
          status: HttpStatus.BAD_REQUEST,
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    delete body['confirmation'];
    await this.repository.save(this.repository.create(body));
    return { message: 'Account has been created', status: HttpStatus.CREATED };
  }

  async login(body: LoginField) {
    const where = [{ name: body.token }, { email: body.token }];
    const findOne = await this.repository.findOne({ where });
    if (!findOne) {
      throw new HttpException(
        {
          message: 'Username or password inccorect',
          status: HttpStatus.BAD_REQUEST,
        },
        HttpStatus.BAD_REQUEST,
      );
    }

    if (!findOne.checkPassword(body.password, findOne.password)) {
      throw new HttpException(
        {
          message: 'Username or password inccorect',
          status: HttpStatus.BAD_REQUEST,
        },
        HttpStatus.BAD_REQUEST,
      );
    }

    const user = (await this.repository.findOne({
      where,
      select: ['id', 'name', 'email', 'pic', 'role', 'created_at'],
    })) as UserEntity;
    user['is_active'] = true;
    await this.repository.manager.save(user);

    return {
      accessToken: await this.jwtService.signAsync(JSON.stringify(user), {
        secret: fs.readFileSync(join(__dirname, '../../../jwtRS256.key'), {
          encoding: 'utf-8',
        }),
      }),
      status: HttpStatus.OK,
    };
  }

  async destroy(id: number, currentUserRole: number) {
    const findOne = await this.repository.findOne({ where: { id } });
    if (!findOne) {
      throw new HttpException(
        {
          message: 'Account not found',
          status: HttpStatus.BAD_REQUEST,
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    if (currentUserRole === 2) {
      throw new HttpException(
        {
          message: "You don't have this access!",
          status: HttpStatus.BAD_REQUEST,
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    await this.repository.delete(findOne.id);
    return { message: 'Account has been deleted', status: HttpStatus.OK };
  }

  async updated(id: number, body: UpdatedSchema) {
    const findOne = await this.repository.findOne({ where: { id } });
    if (!findOne) {
      throw new HttpException(
        {
          message: 'Account not found',
          status: HttpStatus.BAD_REQUEST,
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    findOne.name = body.name;
    findOne.email = body.email;
    findOne.pic = body.pic;
    findOne.role = body.role as unknown as number;
    await this.repository.save(findOne);
    return { message: 'Account has been updated', status: HttpStatus.OK };
  }

  async logout(id: number) {
    const findOne = (await this.repository.findOne({
      where: { id },
    })) as UserEntity;
    findOne.is_active = false;
    await this.repository.manager.save(findOne);
    return { message: 'successfully', status: HttpStatus.OK };
  }
}
