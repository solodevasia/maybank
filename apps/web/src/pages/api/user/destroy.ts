import { NextApiRequest, NextApiResponse } from "next";

export default async function handler(req: NextApiRequest, res: NextApiResponse) {
    const result = await fetch(`http://localhost:4000/api/v1/user/${req.url?.split('user=')[1]}`,{
        method: 'DELETE',
        headers: {'Authorization': `Bearer ${req.headers.cookie?.split('token=').join('')}`}
    }).then(async(resp) => await resp.json())
    res.status(result.status).json(result)
}