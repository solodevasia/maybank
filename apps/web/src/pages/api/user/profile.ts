import { NextApiRequest, NextApiResponse } from "next";

export default async function handler(req: NextApiRequest, res: NextApiResponse) {
    const result = await fetch('http://localhost:4000/api/v1/user/profile', {
        method: 'GET',
        headers: {'Authorization': `Bearer ${req.headers.cookie?.split('token=').join('')}`}
    }).then(async(resp) => await resp.json())
    res.status(result.status).json(result)
}