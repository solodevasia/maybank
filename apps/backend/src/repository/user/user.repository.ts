import { QueryUserList } from '@maybank/dto/user.dto';
import UserEntity from '@maybank/entities/user/user.entity';
import { HttpException, HttpStatus, Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import _ from 'lodash';

@Injectable()
export default class UserRepository {
  constructor(
    @InjectRepository(UserEntity)
    private readonly repository: Repository<UserEntity>,
  ) {}

  async findOne(id: number) {
    const result = await this.repository.findOne({
      where: { id },
      select: ['id', 'name', 'email', 'pic', 'role', 'created_at', 'is_active'],
    });
    if (!result) {
      throw new HttpException(
        {
          message: false,
          status: HttpStatus.BAD_REQUEST,
        },
        HttpStatus.BAD_REQUEST,
      );
    }
    return { result, status: HttpStatus.OK };
  }

  async find(query: QueryUserList) {
    let page = Number(query.page || 1),
      limit = Number(query.limit || 10);
    const result = await this.repository.find({
      where: {
        ...(_.omit(query, ['page', 'limit']) as unknown as UserEntity),
      },
      select: ['id', 'name', 'email', 'pic', 'role', 'created_at', 'is_active'],
      skip: limit * page - limit || 0,
      take: limit,
    });
    const count = await this.repository.count();
    const pageSize = Math.max(Math.floor(count / limit), 1);
    return {
      result,
      page,
      nextUrl: page !== pageSize ? `/user?page=${page + 1}` : null,
      previousUrl: page !== 1 ? `/user?page=${page - 1}` : null,
      pageSize,
      count,
      status: HttpStatus.OK,
    };
  }
}
