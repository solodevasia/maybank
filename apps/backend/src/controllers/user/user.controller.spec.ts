import database from '@maybank/configuration/database';
import UserEntity from '@maybank/entities/user/user.entity';
import UserModule from '@maybank/modules/user.module';
import { HttpStatus, INestApplication } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import { getRepositoryToken, TypeOrmModule } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import supertest from 'supertest';
import fs from 'fs';
import { join } from 'path';
import { JwtModule, JwtService } from '@nestjs/jwt';
import { LoginField, UserRegisterField } from '@maybank/dto/user.dto';
import { faker } from '@faker-js/faker/.';

describe('UserController', () => {
  let app: INestApplication;
  let repository: Repository<UserEntity>;
  let jwtService: JwtService;

  beforeEach(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [
        JwtModule.register({
          global: true,
          secret: fs.readFileSync(join(__dirname, '../../../jwtRS256.key'), {
            encoding: 'utf-8',
          }),
          signOptions: { expiresIn: '60s' },
        }),
        TypeOrmModule.forRoot(database),
        UserModule,
      ],
    }).compile();

    app = moduleRef.createNestApplication();
    repository = moduleRef.get<Repository<UserEntity>>(
      getRepositoryToken(UserEntity),
    );
    jwtService = moduleRef.get(JwtService);
    await app.init();
  });

  it('should to be defined', () => expect(app).toBeDefined());

  it('render correctly', () => expect(app.getHttpServer()).toMatchSnapshot());

  it('should register an account', async () =>
    await supertest(app.getHttpServer())
      .post('/user')
      .set('Content-Type', 'application/json')
      .send({
        name:
          (await repository.count()) <= 2 ? 'admin' : faker.internet.username(),
        email:
          (await repository.count()) <= 2
            ? 'admin@admin.com'
            : faker.internet.email(),
        pic: faker.commerce.productMaterial(),
        role: (await repository.count()) <= 2 ? 1 : 2,
        password: '12345678',
        confirmation: '12345678',
      } as Partial<UserRegisterField>)
      .expect(HttpStatus.CREATED)
      .expect({
        message: 'Account has been created',
        status: HttpStatus.CREATED,
      }));

  it("should register an account but password don't match", async () =>
    await supertest(app.getHttpServer())
      .post('/user')
      .set('Content-Type', 'application/json')
      .send({
        name: faker.internet.username(),
        email: faker.internet.email(),
        password: 'password',
        confirmation: 'passwodwqdwqrd',
      } as Partial<UserRegisterField>)
      .expect(HttpStatus.BAD_REQUEST)
      .expect({
        message: "Password don't match, please check again",
        status: HttpStatus.BAD_REQUEST,
      }));

  it('should register an account but account already exists', async () => {
    const find = await repository.find({
      take: 1,
    });
    await supertest(app.getHttpServer())
      .post('/user')
      .set('Content-Type', 'application/json')
      .send({
        name: find[0].name,
        email: faker.internet.email(),
        password: 'password',
        confirmation: 'password',
      } as Partial<UserRegisterField>)
      .expect(HttpStatus.BAD_REQUEST)
      .expect({
        message: 'Username or email already exists',
        status: HttpStatus.BAD_REQUEST,
      });
  });

  it('should call api "/login/access"', async () => {
    const find = await repository.find({
      take: 1,
      order: { created_at: 'DESC' },
    });
    await supertest(app.getHttpServer())
      .post('/user/login/access')
      .set('Content-Type', 'application/json')
      .send({
        token: find[0].name,
        password: '12345678',
      } as Partial<LoginField>)
      .expect(HttpStatus.OK)
      .then((res) => {
        fs.writeFileSync(
          join(__dirname, '../../../folder/token.txt'),
          res.body.accessToken,
        );
        expect(res.body).toEqual({
          accessToken: res.body.accessToken,
          status: HttpStatus.OK,
        });
      });
  });

  it('should call api "/login/access" but account not found', async () =>
    await supertest(app.getHttpServer())
      .post('/user/login/access')
      .set('Content-Type', 'application/json')
      .send({
        token: 'dwqdqwdq',
        password: 'password',
      } as Partial<LoginField>)
      .expect(HttpStatus.BAD_REQUEST)
      .expect({
        message: 'Username or password inccorect',
        status: HttpStatus.BAD_REQUEST,
      }));

  it('should call api "/login/access" but invalid password', async () => {
    const find = await repository.find({ take: 1 });
    await supertest(app.getHttpServer())
      .post('/user/login/access')
      .set('Content-Type', 'application/json')
      .send({
        token: find[0].name,
        password: 'passwdwqdwqkmkord',
      } as Partial<LoginField>)
      .expect(HttpStatus.BAD_REQUEST)
      .expect({
        message: 'Username or password inccorect',
        status: HttpStatus.BAD_REQUEST,
      });
  });

  it('should call api "/user"', async () =>
    await supertest(app.getHttpServer())
      .get('/user')
      .set('Content-Type', 'application/json')
      .expect(HttpStatus.UNAUTHORIZED));

  it('should call api "/user"', async () =>
    await supertest(app.getHttpServer())
      .get('/user')
      .set('Content-Type', 'application/json')
      .set('Authorization', `Bearer dqwdq`)
      .expect(HttpStatus.UNAUTHORIZED));

  if (
    fs.readFileSync(join(__dirname, '../../../folder/token.txt'), {
      encoding: 'utf-8',
    })
  ) {
    const token = fs.readFileSync(
      join(__dirname, '../../../folder/token.txt'),
      { encoding: 'utf-8' },
    );

    it('should call api destroy "/user/:id"', async () => {
      const admin = await repository.findOne({ where: { role: 1 } });
      const findDestroy = await repository.find({
        where: { role: 2 },
        order: { created_at: 'asc' },
      });
      if ((await repository.count()) >= 8) {
        await supertest(app.getHttpServer())
          .delete(`/user/${findDestroy[0].id}`)
          .set('Content-Type', 'application/json')
          .set(
            'Authorization',
            `Bearer ${await jwtService.signAsync(JSON.stringify(admin), {
              secret: fs.readFileSync(
                join(__dirname, '../../../jwtRS256.key'),
                {
                  encoding: 'utf-8',
                },
              ),
            })}`,
          )
          .expect(HttpStatus.OK)
          .expect({
            message: 'Account has been deleted',
            status: HttpStatus.OK,
          });
      }
    });

    it('should call api destroy "/user/:id" but account not found', async () => {
      const admin = await repository.findOne({ where: { role: 1 } });

      await supertest(app.getHttpServer())
        .delete(`/user/092821`)
        .set('Content-Type', 'application/json')
        .set(
          'Authorization',
          `Bearer ${await jwtService.signAsync(JSON.stringify(admin), {
            secret: fs.readFileSync(join(__dirname, '../../../jwtRS256.key'), {
              encoding: 'utf-8',
            }),
          })}`,
        )
        .expect(HttpStatus.BAD_REQUEST)
        .expect({
          message: 'Account not found',
          status: HttpStatus.BAD_REQUEST,
        });
    });

    it('should call api destroy "/user/:id" but not admin', async () => {
      const findDestroy = await repository.find({
        where: { role: 2 },
        order: { created_at: 'asc' },
      });
      await supertest(app.getHttpServer())
        .delete(`/user/${findDestroy[0].id}`)
        .set('Content-Type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .expect(HttpStatus.BAD_REQUEST)
        .expect({
          message: "You don't have this access!",
          status: HttpStatus.BAD_REQUEST,
        });
    });

    it('should call api "/user/:id"', async () => {
      const user = (await jwtService.verifyAsync(token, {
        secret: fs.readFileSync(join(__dirname, '../../../jwtRS256.key'), {
          encoding: 'utf-8',
        }),
      })) as UserEntity;
      await supertest(app.getHttpServer())
        .put(`/user/${user.id}`)
        .set('Content-Type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .send({
          name:
            user.name === 'admin'
              ? user.name
              : `updated - ${faker.internet.username()}`,
          email:
            user.email === 'admin@admin.com'
              ? user.email
              : `updated${faker.internet.email()}`,
        })
        .expect(HttpStatus.OK)
        .expect({ message: 'Account has been updated', status: HttpStatus.OK });
    });

    it('should call api "/user/:id" account not found', async () => {
      await supertest(app.getHttpServer())
        .put(`/user/092912`)
        .set('Content-Type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .send({
          name: `updated - ${faker.internet.username()}`,
          email: `updated${faker.internet.email()}`,
        })
        .expect(HttpStatus.BAD_REQUEST)
        .expect({
          message: 'Account not found',
          status: HttpStatus.BAD_REQUEST,
        });
    });

    it('should call api "/user/logout"', async () =>
      await supertest(app.getHttpServer())
        .post('/user/logout')
        .set('Content-Type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .expect(HttpStatus.OK)
        .expect({ message: 'successfully', status: HttpStatus.OK }));

    it('should call api "/user"', async () =>
      await supertest(app.getHttpServer())
        .get('/user')
        .set('Content-Type', 'application/json')
        .set('Authorization', `Bearer ${token}`)
        .expect(HttpStatus.OK)
        .then((res) =>
          expect(res.body).toEqual({
            result: res.body.result,
            page: 1,
            previousUrl: null,
            nextUrl: res.body.pageSize >= 2 ? '/user?page=2' : null,
            pageSize: res.body.pageSize,
            count: res.body.count,
            status: HttpStatus.OK,
          }),
        ));

    it('should call api "/user/profile"', async () =>
      await supertest(app.getHttpServer())
        .get('/user/profile')
        .set('Authorization', `Bearer ${token}`)
        .set('Content-Type', 'application/json')
        .expect(HttpStatus.OK)
        .then((res) =>
          expect(res.body).toEqual({
            result: res.body.result,
            status: HttpStatus.OK,
          }),
        ));

    if (
      fs.readFileSync(join(__dirname, '../../../folder/create.txt'), {
        encoding: 'utf-8',
      })
    ) {
      const user = JSON.parse(
        fs.readFileSync(join(__dirname, '../../../folder/create.txt'), {
          encoding: 'utf-8',
        }),
      );

      it('should call api "/user/:id"', async () =>
        await supertest(app.getHttpServer())
          .get(`/user/${user.id}`)
          .set('Content-Type', 'application/json')
          .set('Authorization', `Bearer ${token}`)
          .expect(HttpStatus.OK)
          .then((res) =>
            expect(res.body).toEqual({
              result: res.body.result,
              status: HttpStatus.OK,
            }),
          ));

      it('should call api "/user/:id" not found', async () =>
        await supertest(app.getHttpServer())
          .get(`/user/dqwdwq`)
          .set('Content-Type', 'application/json')
          .set('Authorization', `Bearer ${token}`)
          .expect(HttpStatus.BAD_REQUEST));
    }

    if (
      fs.readFileSync(join(__dirname, '../../../folder/pageSize.txt'), {
        encoding: 'utf-8',
      })
    ) {
      const pageSize = fs.readFileSync(
        join(__dirname, '../../../folder/pageSize.txt'),
        {
          encoding: 'utf-8',
        },
      );

      if (Number(pageSize) >= 3) {
        it('should call api "/user?page=2"', async () =>
          await supertest(app.getHttpServer())
            .get('/user')
            .set('Content-Type', 'application/json')
            .set('Authorization', `Bearer ${token}`)
            .query({ page: 2 })
            .expect(HttpStatus.OK)
            .then((res) => {
              expect(res.body).toEqual({
                result: res.body.result,
                page: 2,
                previousUrl: '/user?page=1',
                nextUrl: '/user?page=3',
                pageSize: res.body.pageSize,
                count: res.body.count,
                status: HttpStatus.OK,
              });
            }));
      }

      it('should call api "/user?page=lastPage"', async () =>
        await supertest(app.getHttpServer())
          .get(`/user?page=${pageSize}`)
          .set('Content-Type', 'application/json')
          .set('Authorization', `Bearer ${token}`)
          .expect(HttpStatus.OK)
          .then(async (res) =>
            expect(res.body).toEqual({
              result: res.body.result,
              page: Number(pageSize),
              previousUrl:
                Number(pageSize) > 1
                  ? `/user?page=${Number(pageSize) - 1}`
                  : null,
              nextUrl: null,
              pageSize: res.body.pageSize,
              count: res.body.count,
              status: HttpStatus.OK,
            }),
          ));
    }
  }

  afterEach(async () => {
    fs.writeFileSync(
      join(__dirname, '../../../folder/pageSize.txt'),
      `${Math.floor((await repository.count()) / 10) || 1}`,
    );
  });
});
