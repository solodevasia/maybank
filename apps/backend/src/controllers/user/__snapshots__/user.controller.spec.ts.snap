// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserController render correctly 1`] = `
Server {
  "_connections": 0,
  "_events": {
    "connection": [Function],
    "listening": [Function],
    "request": [Function],
  },
  "_eventsCount": 3,
  "_handle": null,
  "_listeningId": 1,
  "_maxListeners": undefined,
  "_unref": false,
  "_usingWorkers": false,
  "_workers": [],
  "allowHalfOpen": true,
  "connectionsCheckingInterval": 30000,
  "headersTimeout": 60000,
  "highWaterMark": 65536,
  "httpAllowHalfOpen": false,
  "insecureHTTPParser": undefined,
  "joinDuplicateHeaders": undefined,
  "keepAlive": false,
  "keepAliveInitialDelay": 0,
  "keepAliveTimeout": 5000,
  "maxHeaderSize": undefined,
  "maxHeadersCount": null,
  "maxRequestsPerSocket": 0,
  "noDelay": true,
  "pauseOnConnect": false,
  "rejectNonStandardBodyWrites": false,
  "requestTimeout": 300000,
  "requireHostHeader": true,
  "timeout": 0,
  Symbol(IncomingMessage): [Function],
  Symbol(ServerResponse): [Function],
  Symbol(shapeMode): false,
  Symbol(kCapture): false,
  Symbol(async_id_symbol): -1,
  Symbol(kUniqueHeaders): null,
}
`;
